sdk/storage/inc/blobs/blob_options.hpp

```diff
@@ -191,178 +190,178 @@ namespace Azure { namespace Storage { namespace Blobs {
+    Azure::Core::Nullable<std::string> IfModifiedSince;
+    Azure::Core::Nullable<std::string> IfUnmodifiedSince;
+    Azure::Core::Nullable<std::string> IfMatch;
+    Azure::Core::Nullable<std::string> IfNoneMatch;
   };
 
   struct CreateAppendBlobOptions
   {
     Azure::Core::Context Context;
     BlobHttpHeaders Properties;
     std::map<std::string, std::string> Metadata;
-    std::string IfModifiedSince;
-    std::string IfUnmodifiedSince;
-    std::string IfMatch;
-    std::string IfNoneMatch;
+    Azure::Core::Nullable<std::string> IfModifiedSince;
+    Azure::Core::Nullable<std::string> IfUnmodifiedSince;
+    Azure::Core::Nullable<std::string> IfMatch;
+    Azure::Core::Nullable<std::string> IfNoneMatch;
   };
 
   struct AppendBlockOptions
   {
     Azure::Core::Context Context;
-    std::string ContentMD5;
-    std::string ContentCRC64;
-    std::string LeaseId;
-    uint64_t MaxSize = std::numeric_limits<uint64_t>::max();
-    uint64_t AppendPosition = std::numeric_limits<uint64_t>::max();
-    std::string IfModifiedSince;
-    std::string IfUnmodifiedSince;
-    std::string IfMatch;
-    std::string IfNoneMatch;
+    Azure::Core::Nullable<std::string> ContentMD5;
+    Azure::Core::Nullable<std::string> ContentCRC64;
+    Azure::Core::Nullable<std::string> LeaseId;
+    Azure::Core::Nullable<int64_t> MaxSize;
+    Azure::Core::Nullable<int64_t> AppendPosition;
+    Azure::Core::Nullable<std::string> IfModifiedSince;
+    Azure::Core::Nullable<std::string> IfUnmodifiedSince;
+    Azure::Core::Nullable<std::string> IfMatch;
+    Azure::Core::Nullable<std::string> IfNoneMatch;
   };
 
   struct AppendBlockFromUriOptions
   {
     Azure::Core::Context Context;
-    uint64_t SourceOffset = std::numeric_limits<uint64_t>::max();
-    uint64_t SourceLength = 0;
-    std::string ContentMD5;
-    std::string ContentCRC64;
-    std::string LeaseId;
-    uint64_t MaxSize = std::numeric_limits<uint64_t>::max();
-    uint64_t AppendPosition = std::numeric_limits<uint64_t>::max();
-    std::string IfModifiedSince;
-    std::string IfUnmodifiedSince;
-    std::string IfMatch;
-    std::string IfNoneMatch;
+    Azure::Core::Nullable<int64_t> SourceOffset;
+    Azure::Core::Nullable<int64_t> SourceLength;
+    Azure::Core::Nullable<std::string> ContentMD5;
+    Azure::Core::Nullable<std::string> ContentCRC64;
+    Azure::Core::Nullable<std::string> LeaseId;
+    Azure::Core::Nullable<int64_t> MaxSize;
+    Azure::Core::Nullable<int64_t> AppendPosition;
+    Azure::Core::Nullable<std::string> IfModifiedSince;
+    Azure::Core::Nullable<std::string> IfUnmodifiedSince;
+    Azure::Core::Nullable<std::string> IfMatch;
+    Azure::Core::Nullable<std::string> IfNoneMatch;
   };
 
   struct CreatePageBlobOptions
   {
     Azure::Core::Context Context;
-    uint64_t SequenceNumber = 0;
+    Azure::Core::Nullable<int64_t> SequenceNumber = 0;
     BlobHttpHeaders Properties;
     std::map<std::string, std::string> Metadata;
```

Comment:
Why not `Nullable`?

Reply:
it's already nullable

Reply:
Making a `std::map` nullable? What's the point of doing that?

Reply:
If it was nullable, consider these two situations

1. User doesn't set metadata at all.
2. User sets an empty `std::map` to metadata

What's the difference?

Reply:
If I were a user of storage sdk, I would get confused.

Reply:
Let's think about the original intention of `Nullable`. It was designed to denote the absence of a parameter. For example, for integer parameter, all possible values have their own semantic meanings, and we want another semantic, which is this integer parameter isn't provided or specified. So we need `Nullable` here.

But for metadata, a `std::map` is already able to express all the semantics we want to support. The user provides one entry, or two entries, or not provides at all. Do we need another semantic here? I don't think so. Then what is `Nullable` for here?

Reply:
I don't consider `std::map` any different with `std::string` when it comes to if `Nullable` is applicable. Customer would also set an `IfMatch` to an empty string and we are not sending that header as well. So if `IfMatch` is `Nullable`, a `Metadata` should also be Nullable.

Reply:
No, they're different. `IfMatch` corresponds to one HTTP header. metadata corresponds to a set of HTTP headers. It's not totally reasonable to compare them.

Reply:
Say, if we have a struct `ConditionalHeaders` and it has four members, `IfMatch`, `IfNoneMatch`, `IfModifiedSince`, `IfUnmodifiedSince`. This is a reasonable assumption because both of us ever wanted to have such a struct.

You want to make all of those four members Nullable, or want to make `ConditionalHeaders` itself nullable?

I think it's reasonable to compare `ConditionalHeaders` with metadata. Because they are both a set of HTTP headers.

Reply:
https://github.com/Azure/azure-sdk-for-cpp/issues/232